File: leetcode/recursion + memo/213. House Robber II/js/main.js
// Time O(N) + O(N)
// Space O(N) + O(N)
const rob = nums => {
  if (!nums.length) {
    return 0;
  }

  if (nums.length === 1) {
    return nums[0];
  }

  if (nums.length === 2) {
    return Math.max(nums[0], nums[1]);
  }

  let ans1 = getMax(0, nums.length - 2);
  let ans2 = getMax(1, nums.length - 1);

  return Math.max(ans1, ans2);

  function getMax(start, end) {
    let dp = [];

    dp[start] = nums[start];
    dp[start + 1] = nums[start + 1];

    for (let i = start + 2; i <= end; i++) {
      let max = dp[i - 3] ? Math.max(dp[i - 2], dp[i - 3]) : dp[i - 2] || 0;

      dp[i] = max + nums[i];
    }

    return Math.max(dp[end], dp[end - 1]);
  }
};

File: leetcode/recursion + memo/213. House Robber II/py/main.py
# Time O(N)
# Space O(N)
class Solution:
    def rob(self, nums):
        N = len(nums)
        dp = [[-1] * 2 for _ in range(N)]

        def dfs(index, isFirst):
            if index >= N:
                return 0

            if index == N - 1:
                return 0 if isFirst else nums[N - 1]

            stepId = 0 if isFirst else 1

            if dp[index][stepId] != -1:
                return dp[index][stepId]

            res = 0

            if index == 0:
                # Consider two cases if we start the search from the first index and from the second index
                res = max(res, nums[index] + dfs(index + 2, True), dfs(index + 1, False))
            else:
                # Can take the current value or can go forward
                res = max(res, nums[index] + dfs(index + 2, isFirst), dfs(index + 1, isFirst))

            dp[index][stepId] = res

            return res

        return dfs(0, False)
    